### PR TITLE
Apascual/30094 test mixtral decoder against hf

### LINKED ISSUE #30094: `test_mixtral_decoder.py`
https://github.com/tenstorrent/tt-metal/blob/bf338f0844e710e58de614b919f4e959b6f133a1/models/tt_transformers/tests/mixtral/test_mixtral_decoder.py#L10

### PR DESCRIPTION
**Pull request by adrian-pascual-bernal**
### Ticket
Closes [#30094 ](https://github.com/tenstorrent/tt-metal/issues/30094)

### Problem description
The references for Mixtral tests are being changed to HF.

### What's changed
The decoder block has been changed to HF, some additional code have had to be changed:

 * Rotary encoding: HF version has been used instead.
 * Causal masking: HF version has also been used, only for the prefill tests.

HF config loading has been refactored into a new utilities module.

### Checklist

- [x] [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/runs/20817652671)

#### Model tests

- [x] [(T3K) T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/runs/20817628056)

### CODE REVIEW

File: models/tt_transformers/tests/mixtral/test_mixtral_decoder.py [hunk 1]
```diff
@@ -52,11 +64,14 @@ def test_mixtral_decoder_inference(t3k_mesh_device, reset_seeds, batch, device_p
...
         raise ValueError(f"Batch size {batch} not supported")
 
+    hf_config = load_hf_mixtral_config()
     model_args = ModelArgs(t3k_mesh_device, max_seq_len=max_seq_len, max_batch_size=batch)
+    layer_idx = 0
     state_dict = model_args.load_state_dict()
-    partial_state_dict = {k[9:]: v for k, v in state_dict.items() if (k.startswith("layers.0."))}
-    reference_model = TransformerBlock(args=model_args)
+    partial_state_dict = {k[9:]: v for k, v in state_dict.items() if (k.startswith(f"layers.{layer_idx}."))}
+    reference_model = MixtralDecoderLayer(hf_config, layer_idx)
```
Comment (maksim-tsishkouski-epam) [inline models/tt_transformers/tests/mixtral/test_mixtral_decoder.py:73]: I'm ok with the current solution, just curious if it's possible to use
model_args.reference_decoder as it's done in test_decoder tests?
https://github.com/tenstorrent/tt-metal/blob/83e2c59e032a8c470b30d80b666f294588d9f219/models/tt_transformers/tests/test_decoder.py#L73

**Review by maksim-tsishkouski-epam — APPROVED** [review]

File: models/tt_transformers/tests/mixtral/test_mixtral_decoder.py (hunk 1 above)
Comment (adrian-pascual-bernal) [reply models/tt_transformers/tests/mixtral/test_mixtral_decoder.py:73]: TBH I wasn't aware of such method when I wrote this code, I'm going to explore this option, I think it may be a good idea to use it to keep the code more consistent

File: models/tt_transformers/tests/mixtral/test_mixtral_decoder_prefill.py [hunk 2]
```diff
@@ -45,20 +63,21 @@ def test_mixtral_decoder_inference(t3k_mesh_device, reset_seeds, batch, device_p
...
+    hf_config = load_hf_mixtral_config()
     model_args = ModelArgs(t3k_mesh_device, max_seq_len=max_seq_len, max_batch_size=batch)
     model_args.n_layers = 1
 
     t3k_mesh_device.disable_and_clear_program_cache()
 
     state_dict = model_args.load_state_dict()
-
-    first_layer_prefix = model_args.get_state_dict_prefix("TransformerBlock", 0)
+    layer_idx = 0
```
Comment (gwangTT) [inline models/tt_transformers/tests/mixtral/test_mixtral_decoder_prefill.py]: I like this change better than the previous magic constant. 

nitpick: it may be even nicer to have this parameter be part of pytest mark parameterize list.

File: models/tt_transformers/tests/mixtral/test_mixtral_decoder.py [hunk 3]
```diff
@@ -52,11 +64,14 @@ def test_mixtral_decoder_inference(t3k_mesh_device, reset_seeds, batch, device_p
     else:
         raise ValueError(f"Batch size {batch} not supported")
 
+    hf_config = load_hf_mixtral_config()
     model_args = ModelArgs(t3k_mesh_device, max_seq_len=max_seq_len, max_batch_size=batch)
+    layer_idx = 0
```
Comment (gwangTT) [inline models/tt_transformers/tests/mixtral/test_mixtral_decoder.py]: I like this change better than the previous magic constant.

nitpick: it may be even nicer to have this parameter be part of pytest mark parameterize list.

**Review by gwangTT — APPROVED** [review]
Three tests in one go! Thanks, Adrian!

**Review by gwangTT — COMMENTED** [review]
I see that the tests in this PR are all part of `tests/scripts/t3000/run_t3000_unit_tests.sh` (run in the t3000 unit test CI pipeline). 

Can you please add a CI run with that pipeline and link it in the Checklist section of the PR description?

File: models/tt_transformers/tests/mixtral/test_mixtral_decoder.py (hunk 3 above)
Comment (adrian-pascual-bernal) [reply models/tt_transformers/tests/mixtral/test_mixtral_decoder.py]: I can modify it to be part of the parametrize list, I agree on it to be nicer.

**Comment by adrian-pascual-bernal** [comment]
> I see that the tests in this PR are all part of `tests/scripts/t3000/run_t3000_unit_tests.sh` (run in the t3000 unit test CI pipeline).
> 
> Can you please add a CI run with that pipeline and link it in the Checklist section of the PR description?

Do you mean a different pipeline from [this one](https://github.com/tenstorrent/tt-metal/actions/runs/20578682636)?

**Comment by gwangTT** [comment]
> > I see that the tests in this PR are all part of `tests/scripts/t3000/run_t3000_unit_tests.sh` (run in the t3000 unit test CI pipeline).
> > Can you please add a CI run with that pipeline and link it in the Checklist section of the PR description?
> 
> Do you mean a different pipeline from [this one](https://github.com/tenstorrent/tt-metal/actions/runs/20578682636)?

Oh, did I misread the checklist in the description? That is the one. Though you may want to run it again after resolving the conflicts.
